src/test/func/jemalloc/jemalloc.cc

```diff
@@ -0,0 +1,393 @@
+#include <functional>
+#include <stdio.h>
+#include <test/setup.h>
+
+#define SNMALLOC_NAME_MANGLE(a) our_##a
+#undef SNMALLOC_NO_REALLOCARRAY
+#undef SNMALLOC_NO_REALLOCARR
+#define SNMALLOC_BOOTSTRAP_ALLOCATOR
+#include "../../../override/malloc.cc"
+
+#if __has_include(<malloc_np.h>)
+#  include <malloc_np.h>
+#endif
+
+#ifdef __FreeBSD__
+/**
+ * Enable testing against the versions that we get from libc or elsewhere.
+ * Enabled by default on FreeBSD where all of the jemalloc functions are
+ * exported from libc.
+ */
+#  define TEST_JEMALLOC_MALLOCX
+#endif
+
+#define OUR_MALLOCX_LG_ALIGN(la) (static_cast<int>(la))
+#define OUR_MALLOCX_ZERO (one_at_bit<int>(6))
+
+#define OUR_ALLOCM_NO_MOVE (one_at_bit<int>(7))
+
+#define OUR_ALLOCM_SUCCESS 0
+#define OUR_ALLOCM_ERR_OOM 1
+#define OUR_ALLOCM_ERR_NOT_MOVED 2
+
+#ifndef MALLOCX_LG_ALIGN
+#  define MALLOCX_LG_ALIGN(la) OUR_MALLOCX_LG_ALIGN(la)
+#endif
+#ifndef MALLOCX_ZERO
+#  define MALLOCX_ZERO OUR_MALLOCX_ZERO
+#endif
+
+#ifndef ALLOCM_LG_ALIGN
+#  define ALLOCM_LG_ALIGN(la) OUR_MALLOCX_LG_ALIGN(la)
+#endif
+#ifndef ALLOCM_ZERO
+#  define ALLOCM_ZERO OUR_MALLOCX_ZERO
+#endif
+#ifndef ALLOCM_NO_MOVE
+#  define ALLOCM_NO_MOVE OUR_ALLOCM_NO_MOVE
+#endif
+#ifndef ALLOCM_SUCCESS
+#  define ALLOCM_SUCCESS OUR_ALLOCM_SUCCESS
+#endif
+#ifndef ALLOCM_ERR_OOM
+#  define ALLOCM_ERR_OOM OUR_ALLOCM_ERR_OOM
+#endif
+#ifndef ALLOCM_ERR_NOT_MOVED
+#  define ALLOCM_ERR_NOT_MOVED OUR_ALLOCM_ERR_NOT_MOVED
+#endif
+
+#ifdef _MSC_VER
+#  define __PRETTY_FUNCTION__ __FUNCSIG__
+#endif
+
+using namespace snmalloc;
+using namespace snmalloc::bits;
+
+namespace
+{
+  /**
+   * Test whether the MALLOCX_LG_ALIGN macro is defined correctly.  This test
+   * will pass trivially if we don't have the malloc_np.h header from
+   * jemalloc, but at least the FreeBSD action runners in CI do have this
+   * header.
+   */
+  template<int Size>
+  void check_lg_align_macro()
+  {
+    static_assert(
+      OUR_MALLOCX_LG_ALIGN(Size) == MALLOCX_LG_ALIGN(Size),
+      "Our definition of MALLOCX_LG_ALIGN is wrong");
+    static_assert(
+      OUR_MALLOCX_LG_ALIGN(Size) == ALLOCM_LG_ALIGN(Size),
+      "Our definition of ALLOCM_LG_ALIGN is wrong");
+    static_assert(
+      JEMallocFlags(Size).log2align() == Size, "Out log2 align mask is wrong");
+    if constexpr (Size > 0)
+    {
+      check_lg_align_macro<Size - 1>();
+    }
+  }
+
+  /**
+   * The name of the function under test.  This is set in the START_TEST macro
+   * and used for error reporting in EXPECT.
+   */
+  const char* function = nullptr;
+
+  /**
+   * Log that the test started.
+   */
+#define START_TEST(msg) \
+  function = __PRETTY_FUNCTION__; \
+  fprintf(stderr, "Starting test: " msg "\n");
+
+  /**
+   * An assertion that fires even in debug builds.  Uses the value set by
+   * START_TEST.
+   */
+#define EXPECT(x, msg, ...) \
```

Comment:
Could we use `SNMALLOC_CHECK`
https://github.com/microsoft/snmalloc/blob/main/src/ds/defines.h#L125

Reply:
Can you improve SNMALLOC_CHECK?  It is only used in tests. I just would like to not have a proliferation of different bits for doing basically the same thing.

Reply:
I don't think so easily.  `SNMALLOC_CHECK` calls `Pal::error`, which carefully avoids depending on `malloc`.  The `EXPECT` macro doesn't because it is never used in a situation where we expect `malloc` to be too broken for `fprintf` to work.  If we want to use `SNMALLOC_CHECK` for tests where the allocator under test is the process' `malloc` implementation then we can't make it do this.

Once we can depend on a C++ standard library that has `std::format` and `std::source_location` then we can make much nicer versions of both of these and so I'd prefer to limit the scope of a macro that uses `printf` format strings until we're willing to depend on a C++20 stdlib for the tests (even on platforms where we have a C++20 compiler, we don't always have a C++20 stdlib at the moment).  I could write a C++20 version that used `constexpr` if things to avoid memory allocation if the message didn't use use any format arguments, which would give the right behaviour in both cases, but it would break CI and a bunch of our not-tested-in-CI platforms.

I'm happy to do that in another PR that we don't merge until we're happy that platforms we care about have a sufficiently modern stdlib, if you like.

Reply:
Okay. Makes sense.  Lets perhaps look at lifting something into `tests/utils.h` or something of that nature, so we can share across tests. (As part of a future PR.)